plugins/cachekey/cachekey.cc

```diff
@@ -437,6 +437,61 @@ CacheKey::appendPath(Pattern &pathCapture, Pattern &pathCaptureUri)
   }
 }
 
+template <class T>
+void
+CacheKey::processHeader(const String &name, const ConfigHeaders &config, T &dst,
+                        void (*fun)(const ConfigHeaders &config, const String &name_s, const String &value_s, T &captures))
+{
+  TSMLoc field;
+
+  for (field = TSMimeHdrFieldFind(_buf, _hdrs, name.c_str(), name.size()); field != TS_NULL_MLOC;
+       field = ::nextDuplicate(_buf, _hdrs, field)) {
+    const char *value;
+    int vlen;
+    int count = TSMimeHdrFieldValuesCount(_buf, _hdrs, field);
+
+    for (int i = 0; i < count; ++i) {
+      value = TSMimeHdrFieldValueStringGet(_buf, _hdrs, field, i, &vlen);
+      if (value == nullptr || vlen == 0) {
+        CacheKeyDebug("missing value %d for header %s", i, name.c_str());
+        continue;
+      }
+
+      String value_s(value, vlen);
+      fun(config, name, value_s, dst);
+    }
+  }
+}
+
+template <class T>
+void
+captureWholeHeaders(const ConfigHeaders &config, const String &name, const String &value, T &captures)
+{
+  CacheKeyDebug("processing header %s", name.c_str());
+  if (config.toBeAdded(name)) {
+    String header;
+    header.append(name).append(":").append(value);
+    captures.insert(header);
+    CacheKeyDebug("adding header '%s: %s'", name.c_str(), value.c_str());
+  } else {
+    CacheKeyDebug("failed to find header '%s'", name.c_str());
+  }
+}
+
+template <class T>
+void
+captureFromHeaders(const ConfigHeaders &config, const String &name, const String &value, T &captures)
+{
+  CacheKeyDebug("processing capture from header %s", name.c_str());
+  auto itMp = config.getCaptures().find(name);
+  if (config.getCaptures().end() != itMp) {
+    itMp->second->capture(value, captures);
+    CacheKeyDebug("found capture pattern for header '%s'", name.c_str());
+  } else {
+    CacheKeyDebug("failed to find header '%s'", name.c_str());
```

Comment:
Same comment as above: Will this get too noisy?

Reply:
@d2r please see my comment above

Reply:
OK